### PR TITLE
Improvement on reproducible builds by using Debian snapshots repository as base

### PR DESCRIPTION
**Pull request by accumulator**
Occasionally the AppImage build is not able to be reproducibly built, due to dependencies in the appimage build container becoming unavailable. The quick fix is to upgrade the dependencies in the Dockerfile to newer versions. However, these newer versions of dependencies could affect the binary result of the appimage, resulting in non-reproducibility of older Electrum versions.

This PR implements the idea discussed in https://github.com/spesmilo/electrum/issues/7484

The base image is changed to debian:buster and replaces ubuntu:18.04 (similar age). 

With `contrib/freeze_appimage_distro.sh` an APT `sources.list` file is generated which points to the most recent snapshot timestamp available on `snapshot.debian.org` (e.g. https://snapshot.debian.org/archive/debian/20220810T094023Z/).

Additionally, an APT preferences file is added, setting the Pin-Priority to 1001 to allow for downgrading of packages and forcing all distro dependencies to be taken from the snapshot. The reason for this is that the `ca-certificates` package needed to access `snapshot.debian.org` pulls in a potentially too new libssl.

### CODE REVIEW

File: contrib/freeze_appimage_distro.sh [hunk 1]
```diff
@@ -0,0 +1,33 @@
+#!/bin/sh
```
Comment (SomberNight) [inline contrib/freeze_appimage_distro.sh:1]: Should this file maybe place instead into `contrib/build-linux/appimage/`? and if so, renamed, I guess.
In particular, note that the same original issue also applies to the wine build, hence this kind of solution could be used there as well.

**Review by SomberNight — COMMENTED** [review]
Thanks for this! Looks good overall.

I have a concept question in https://github.com/spesmilo/electrum/issues/7484#issuecomment-1211118678 - see there

File: contrib/freeze_appimage_distro.sh (hunk 1 above)
Comment (accumulator) [reply contrib/freeze_appimage_distro.sh:1]: I don't have a strong preference where the script resides. However maybe we can extend the script to pin, in one go, the timestamp for the appimage AND the wine build (and maybe the android build container as well), which would make contrib/ the obvious location. What do you think?

**Comment by accumulator** [comment]
I changed the build-wine container to use the same approach, to have everything in this branch.

One slight difference is the way I pin the packages. As I need one pkg from sid/unstable in the wine build I pin the bulk to the release name, which needs to be updated if we change to a new debian release. The build properly fails if it's incorrect, so it's not a big issue IMO, but we might want to keep things the same for build-linux/appimage.. No strong opinion on this though.

**Comment by accumulator** [comment]
I guess we need to take the android build along too in this approach? There's a TODO in the Dockerfile we can get rid of ;)

File: contrib/freeze_appimage_distro.sh (hunk 1 above)
Comment (SomberNight) [reply contrib/freeze_appimage_distro.sh]: Sounds good. If the script touches both appimage and wine, than it can be in `contrib/` directly.

**Comment by SomberNight** [comment]
> I guess we need to take the android build along too in this approach? There's a TODO in the Dockerfile we can get rid of ;)

Sure, feel free to have a go at it. The PR is already useful even without that though.

**Comment by accumulator** [comment]
Done. android, appimage and wine all build successful here, although I haven't tested the wine build, and the appimage only very minimally.

**Comment by SomberNight** [comment]
I tried to build from this today but got a failure for the wine build:
```
E: Release file for https://snapshot.debian.org/archive/debian/20220811T031049Z/dists/sid/InRelease is expired (invalid since 1d 11h 56min 39s). Updates for this repository will not be applied.
```
I guess we should avoid using `sid`.


<details>
 <summary>log</summary>

```
+ win1=electrum-4.3.0-17-gdaae6f46c.exe
+ win2=electrum-4.3.0-17-gdaae6f46c-portable.exe
+ win3=electrum-4.3.0-17-gdaae6f46c-setup.exe
+ test -f dist/electrum-4.3.0-17-gdaae6f46c.exe
+ pushd .
~/wspace/electrum ~/wspace/electrum
+ test -f contrib/build-wine/dist/electrum-4.3.0-17-gdaae6f46c.exe
+ ./contrib/build-wine/build.sh
💬 INFO:  Clearing /home/user/wspace/electrum/contrib/build-wine/../../contrib/build-wine/dist...
💬 INFO:  building docker image.
Sending build context to Docker daemon  1.302GB
Step 1/19 : FROM debian:bullseye@sha256:82bab30ed448b8e2509aabe21f40f0607d905b7fd0dec72802627a20274eba55
docker.io/library/debian:bullseye@sha256:82bab30ed448b8e2509aabe21f40f0607d905b7fd0dec72802627a20274eba55: Pulling from library/debian
001c52e26ad5: Pull complete 
Digest: sha256:82bab30ed448b8e2509aabe21f40f0607d905b7fd0dec72802627a20274eba55
Status: Downloaded newer image for debian:bullseye@sha256:82bab30ed448b8e2509aabe21f40f0607d905b7fd0dec72802627a20274eba55
 ---> 07d9246c53a6
Step 2/19 : RUN apt update -qq > /dev/null && apt install -qq --yes --no-install-recommends     ca-certificates
 ---> Running in 18d133c10749

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.


WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

The following additional packages will be installed:
  openssl
The following NEW packages will be installed:
  ca-certificates openssl
debconf: delaying package configuration, since apt-utils is not installed
0 upgraded, 2 newly installed, 0 to remove and 4 not upgraded.
Need to get 1011 kB of archives.
After this operation, 1892 kB of additional disk space will be used.
Selecting previously unselected package openssl.
(Reading database ... 6661 files and directories currently installed.)
Preparing to unpack .../openssl_1.1.1n-0+deb11u3_amd64.deb ...
Unpacking openssl (1.1.1n-0+deb11u3) ...
Selecting previously unselected package ca-certificates.
Preparing to unpack .../ca-certificates_20210119_all.deb ...
Unpacking ca-certificates (20210119) ...
Setting up openssl (1.1.1n-0+deb11u3) ...
Setting up ca-certificates (20210119) ...
debconf: unable to initialize frontend: Dialog
debconf: (TERM is not set, so the dialog frontend is not usable.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (Can't locate Term/ReadLine.pm in @INC (you may need to install the Term::ReadLine module) (@INC contains: /etc/perl /usr/local/lib/x86_64-linux-gnu/perl/5.32.1 /usr/local/share/perl/5.32.1 /usr/lib/x86_64-linux-gnu/perl5/5.32 /usr/share/perl5 /usr/lib/x86_64-linux-gnu/perl-base /usr/lib/x86_64-linux-gnu/perl/5.32 /usr/share/perl/5.32 /usr/local/lib/site_perl) at /usr/share/perl5/Debconf/FrontEnd/Readline.pm line 7.)
debconf: falling back to frontend: Teletype
Updating certificates in /etc/ssl/certs...
129 added, 0 removed; done.
Processing triggers for ca-certificates (20210119) ...
Updating certificates in /etc/ssl/certs...
0 added, 0 removed; done.
Running hooks in /etc/ca-certificates/update.d...
done.
Removing intermediate container 18d133c10749
 ---> 0edc4079829f
Step 3/19 : COPY apt.sources.list /etc/apt/sources.list
 ---> ef6201af26f1
Step 4/19 : COPY apt.preferences /etc/apt/preferences.d/snapshot
 ---> 3873e2016b2b
Step 5/19 : ENV LC_ALL=C.UTF-8 LANG=C.UTF-8
 ---> Running in 6d98cd7d79c4
Removing intermediate container 6d98cd7d79c4
 ---> 5d4a07736c45
Step 6/19 : ENV DEBIAN_FRONTEND=noninteractive
 ---> Running in dca5bc02c1fe
Removing intermediate container dca5bc02c1fe
 ---> 5833d89eb0be
Step 7/19 : RUN dpkg --add-architecture i386 &&     apt-get update -q &&     apt-get install -qy         wget         gnupg2         dirmngr         python3-software-properties         software-properties-common         git         p7zip-full         make         mingw-w64         mingw-w64-tools         win-iconv-mingw-w64-dev         autotools-dev         autoconf         autopoint         libtool         gettext         &&     rm -rf /var/lib/apt/lists/* &&     apt-get autoremove -y &&     apt-get clean
 ---> Running in df62c65e4397
Get:1 https://snapshot.debian.org/archive/debian/20220811T031049Z bullseye InRelease [116 kB]
Get:2 https://snapshot.debian.org/archive/debian/20220811T031049Z sid InRelease [192 kB]
Get:3 https://snapshot.debian.org/archive/debian/20220811T031049Z bullseye/non-free Sources [81.1 kB]
Get:4 https://snapshot.debian.org/archive/debian/20220811T031049Z bullseye/contrib Sources [43.2 kB]
Get:5 https://snapshot.debian.org/archive/debian/20220811T031049Z bullseye/main Sources [8635 kB]
Get:6 https://snapshot.debian.org/archive/debian/20220811T031049Z bullseye/main i386 Packages [8121 kB]
Get:7 https://snapshot.debian.org/archive/debian/20220811T031049Z bullseye/contrib i386 Packages [45.3 kB]
Get:8 https://snapshot.debian.org/archive/debian/20220811T031049Z bullseye/non-free i386 Packages [79.3 kB]
Get:9 https://snapshot.debian.org/archive/debian/20220811T031049Z bullseye/contrib amd64 Packages [50.6 kB]
Get:10 https://snapshot.debian.org/archive/debian/20220811T031049Z bullseye/main amd64 Packages [8182 kB]
Get:11 https://snapshot.debian.org/archive/debian/20220811T031049Z bullseye/non-free amd64 Packages [97.6 kB]
Reading package lists...
E: Release file for https://snapshot.debian.org/archive/debian/20220811T031049Z/dists/sid/InRelease is expired (invalid since 1d 11h 56min 39s). Updates for this repository will not be applied.
The command '/bin/sh -c dpkg --add-architecture i386 &&     apt-get update -q &&     apt-get install -qy         wget         gnupg2         dirmngr         python3-software-properties         software-properties-common         git         p7zip-full         make         mingw-w64         mingw-w64-tools         win-iconv-mingw-w64-dev         autotools-dev         autoconf         autopoint         libtool         gettext         &&     rm -rf /var/lib/apt/lists/* &&     apt-get autoremove -y &&     apt-get clean' returned a non-zero code: 100
```
</details>

I pushed a commit (https://github.com/spesmilo/electrum/pull/7926/commits/e90ff4800c040b50cb19972fba0447a6a35d28bd) where we build the missing package `win-iconv-mingw-w64-dev` ourselves instead of using sid.

File: contrib/freeze_containers_distro.sh [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+echo "deb ${DEBIAN_SNAPSHOT} ${DEBIAN_APPIMAGE_DISTRO} main non-free contrib" >$contrib/build-linux/appimage/apt.sources.list
+echo "deb-src ${DEBIAN_SNAPSHOT} ${DEBIAN_APPIMAGE_DISTRO} main non-free contrib" >>$contrib/build-linux/appimage/apt.sources.list
+
+# build-wine
+echo "deb ${DEBIAN_SNAPSHOT} ${DEBIAN_WINE_DISTRO} main non-free contrib" >$contrib/build-wine/apt.sources.list
+echo "deb-src ${DEBIAN_SNAPSHOT} ${DEBIAN_WINE_DISTRO} main non-free contrib" >>$contrib/build-wine/apt.sources.list
+
+# android
+echo "deb ${DEBIAN_SNAPSHOT} ${DEBIAN_ANDROID_DISTRO} main non-free contrib" >$contrib/build-android/apt.sources.list
+echo "deb-src ${DEBIAN_SNAPSHOT} ${DEBIAN_ANDROID_DISTRO} main non-free contrib" >>$contrib/build-android/apt.sources.list
```
Comment (SomberNight) [inline contrib/freeze_containers_distro.sh:44]: typo in path:
```suggestion
echo "deb ${DEBIAN_SNAPSHOT} ${DEBIAN_ANDROID_DISTRO} main non-free contrib" >$contrib/android/apt.sources.list
echo "deb-src ${DEBIAN_SNAPSHOT} ${DEBIAN_ANDROID_DISTRO} main non-free contrib" >>$contrib/android/apt.sources.list
```

Comment (SomberNight) [reply contrib/freeze_containers_distro.sh]: fixed in force-push https://github.com/spesmilo/electrum/compare/a85e644a763ee85c9e3077efbbfd7b7e1f8b6bc9..0ac49ea8c492a19da4c06d7669c6e2199cfaec38

**Review by SomberNight — APPROVED** [review]
Looks good. Thanks!
I've pushed some fixups to the branch, mainly to have it correctly build on the CI.

**Pull request merged**